crates/hir-expand/src/lib.rs

```diff
@@ -217,8 +220,27 @@ pub struct MacroCallLoc {
 }
 impl InternValueTrivial for MacroCallLoc {}
 
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Ord, PartialOrd)]
+pub struct Macro2Id(pub salsa::InternId);
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Ord, PartialOrd)]
+pub struct MacroRulesId(pub salsa::InternId);
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Ord, PartialOrd)]
+pub struct ProcMacroId(pub salsa::InternId);
+
+/// A macro
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, PartialOrd, Ord)]
+pub enum MacroId {
+    Macro2Id(Macro2Id),
+    MacroRulesId(MacroRulesId),
+    ProcMacroId(ProcMacroId),
+}
+impl_from!(Macro2Id, MacroRulesId, ProcMacroId for MacroId);
+
 #[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
 pub struct MacroDefId {
+    pub id: MacroId,
```

Comment:
Instead of doing this, can we just resolve the `MacroId` to its `MacroDefId` for now via `DefDatabase::macro_def`?

Reply:
Hmm... `MacroDefId` contains an `AstId`. Won't it hurt incrementality?

Reply:
Only if we store it in some query result again which I don't think is the case? (will double check that tomorrow)

Reply:
That means we will have to store `MacroDefId` in body, and then access it during it inference etc. Although macro definitions in bodies are quite rare, so maybe it's okay.

Reply:
Ah we store this in the `ScopeData` right, I missed that. That still shouldn't be a problem though I think, if the Macro's AstId changes we recompute a bunch of base queries anyways so we aren't losing out on incrementality here. (still looking into this more though to see if there is other problems)

Reply:
https://github.com/Veykril/rust-analyzer/commit/883c4b9c9dd3328bd3678546cd2a9b6d31ffa865 is the change changing this back, from what I can tell this should be fine (aside from `ScopeData` becoming bigger, that might be a bit much)

Main reason I don't like the change from your PR here is that it makes the Macro ID stuff even more confusing (it is already a nightmare to grasp imo), I'd love if we could simplify it somehow but that likely should wait for the next salsa either way.

Reply:
Okay, I changed this to store `MacroDefId`.